designs/full-cluster-lifecycle-api-cloudstack.md

```diff
@@ -0,0 +1,81 @@
+
+### Runtime Validation
+
+#### Validator Initialization
+
+* **How to read credential**
+
+  The validator requires cloudstack credential to build cmk. Our current CLI reads credential from env, and has a validation that accept new credential but not allowing existing credential update.
+
+    1. Read credentials from existing secrets which requires customers to add additional secrets. We also need to add a mechanism to prevent customers from update existing secrets.
+    2. (preferred) Read credentials from env like current CLI flow, which could prevent customers from updating credentials and also is an easy way to add new credentials.
+
+
+* **Where to build validator**
+
+    We can use the [validator factory](https://github.com/aws/eks-anywhere/blob/3c1fd0ff732641ed02137213863942403f59c320/pkg/providers/cloudstack/validator_registry.go#L25) to pass validatorRegistry to controller, so the validator is built in every reconcile loop.
+
+We’ll have CloudStackDatacenterReconciler to validate and update corresponding status and failure message.
+
+```
+func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *anywherev1.Cluster) (controller.Result, error) {
+    log = log.WithValues("provider", "cloudstack")
+    clusterSpec, err := c.BuildSpec(ctx, clientutil.NewKubeClient(r.client), cluster)
+    if err != nil {
+        return controller.Result{}, err
+    }
+    
+    return controller.NewPhaseRunner().Register(
+        r.ValidateAndSetEnv,
+        r.ipValidator.ValidateControlPlaneIP,      
+        r.ValidateDatacenterConfig,  
+        r.ValidateMachineConfig,         
+        clusters.CleanupStatusAfterValidate,      
+        r.ReconcileControlPlane,      
+        r.CheckControlPlaneReady,      
+        r.ReconcileCNI,      
+        r.ReconcileWorkers,
+    ).Run(ctx, log, clusterSpec)
+}
+```
+
+* validateAndSetEnv: setting [eksa license](https://github.com/aws/eks-anywhere/blob/3c1fd0ff732641ed02137213863942403f59c320/pkg/providers/cloudstack/cloudstack.go#L395), validate OS, [validate k8s version](https://github.com/aws/eks-anywhere/blob/ed4425dadb19600b4eb446d29b81f5c2441c16f6/pkg/providers/cloudstack/cloudstack.go#L1371),
+  [setDefaultAndValidateControlPlaneHostPort](https://github.com/aws/eks-anywhere/blob/3c1fd0ff732641ed02137213863942403f59c320/pkg/providers/cloudstack/validator.go#L211)
+* ValidateDatacenterConfig checks status and failure message from CloudStackDatacenterReconciler
+* After successful datacenter config validation, ValidateMachineConfig validates run time machine config for each availability zone in datacenter.
```

Comment:
what is `validates run time machine config`?

Reply:
validations that need to talk via cmk